GVFS/GVFS.Service/ProductUpgradeTimer.cs

```diff
@@ -53,70 +52,101 @@ public void Dispose()
-                    return false;
-                }
+                    // The upgrade check always goes against GitHub
+                    GitHubUpgrader productUpgrader = GitHubUpgrader.Create(
+                        this.tracer,
+                        false,
+                        false,
+                        out errorMessage);
 
-                if (!productUpgrader.TryQueryNewestVersion(out newerVersion, out detailedError))
-                {
-                    errorMessage = "Could not fetch new version info. " + detailedError;
-                    return false;
-                }
+                    if (productUpgrader == null)
+                    {
+                        if (errorMessage != null)
+                        {
+                            this.tracer.RelatedError(errorMessage);
+                        }
 
-                if (newerVersion == null)
-                {
-                    // Already up-to-date
-                    // Make sure there a no asset installers remaining in the Downloads directory. This can happen if user
-                    // upgraded by manually downloading and running asset installers.
-                    ProductUpgraderInfo.DeleteAllInstallerDownloads();
-                    errorMessage = null;
-                    return true;
-                }
+                        return;
+                    }
 
-                if (productUpgrader.TryDownloadNewestVersion(out detailedError))
-                {
-                    errorMessage = null;
-                    return true;
+                    InstallerPreRunChecker prerunChecker = new InstallerPreRunChecker(this.tracer, string.Empty);
+                    if (!prerunChecker.TryRunPreUpgradeChecks(out _))
+                    {
+                        return;
+                    }
+
+                    if (!productUpgrader.UpgradeAllowed(out errorMessage))
+                    {
+                        if (errorMessage != null)
+                        {
+                            this.tracer.RelatedError(errorMessage);
+                        }
+
+                        return;
+                    }
+
+                    if (!this.TryQueryForNewerVersion(
+                            activity,
+                            productUpgrader,
+                            out Version newerVersion,
+                            out errorMessage))
+                    {
+                        if (errorMessage != null)
+                        {
+                            this.tracer.RelatedError(errorMessage);
+                        }
+
+                        return;
+                    }
+
+                    ProductUpgraderInfo.RecordHighestAvailableVersion(newerVersion);
                 }
-                else
+                catch (Exception ex) when (
+                    ex is IOException ||
+                    ex is UnauthorizedAccessException ||
+                    ex is NotSupportedException)

```

Comment:
Should we be catching all exception types here?  

It looks like if some other exception was thrown the background thread would die but the service would keep running.